bpo-37085: Expose SocketCAN bcm_msg_head flags

karlding opened the pull request:
Expose the CAN_BCM SocketCAN constants used in the bcm_msg_head struct
flags (provided by <linux/can/bcm.h>) under the socket library.

This adds the following constants with a CAN_BCM prefix:

  * SETTIMER
  * STARTTIMER
  * TX_COUNTEVT
  * TX_ANNOUNCE
  * TX_CP_CAN_ID
  * RX_FILTER_ID
  * RX_CHECK_DLC
  * RX_NO_AUTOTIMER
  * RX_ANNOUNCE_RESUME
  * TX_RESET_MULTI_IDX
  * RX_RTR_FRAME
  * CAN_FD_FRAME

<!-- issue-number: [bpo-37085](https://bugs.python.org/issue37085) -->
https://bugs.python.org/issue37085
<!-- /issue-number -->

karlding committed:
bpo-37085: Expose SocketCAN bcm_msg_head flags

Expose the CAN_BCM SocketCAN constants used in the bcm_msg_head struct
flags (provided by <linux/can/bcm.h>) under the socket library.

This adds the following constants with a CAN_BCM prefix:

  * SETTIMER
  * STARTTIMER
  * TX_COUNTEVT
  * TX_ANNOUNCE
  * TX_CP_CAN_ID
  * RX_FILTER_ID
  * RX_CHECK_DLC
  * RX_NO_AUTOTIMER
  * RX_ANNOUNCE_RESUME
  * TX_RESET_MULTI_IDX
  * RX_RTR_FRAME
  * CAN_FD_FRAME

auvipy reviewed (changes requested):
tests?

taleinat commented on Modules/socketmodule.c:
```diff
@@ -7668,6 +7668,42 @@ PyInit__socket(void)
     PyModule_AddIntConstant(m, "CAN_BCM_RX_STATUS", RX_STATUS);
     PyModule_AddIntConstant(m, "CAN_BCM_RX_TIMEOUT", RX_TIMEOUT);
     PyModule_AddIntConstant(m, "CAN_BCM_RX_CHANGED", RX_CHANGED);
+#ifdef  SETTIMER
```
Why these separate `#ifdef` checks for each constant? Are they not either all available together, or not? More specifically, isn't the `#ifdef HAVE_LINUX_CAN_BCM_H` check enough to ensure their availability?

karlding replied:
This is not necessarily true for all the constants, namely the backwards-compatible introduction of the `CAN_FD_FRAME` flag, which was added in `6f3b911d5f29b98752e5da86a295210c0c4f4e14` as part of the the `4.8.x` kernel series.

The other constants were always present since the SocketCAN drivers were mainlined in `2.6.25`.

Should I just have the `#ifdef` on `CAN_FD_FRAME` then? I had added the other ones for consistency.

taleinat replied:
> Should I just have the #ifdef on CAN_FD_FRAME then? I had added the other ones for consistency.

IMO that would be better, yes. And perhaps add a note about the availability of `CAN_FD_FRAME` in [the docs](https://docs.python.org/3/library/socket.html#socket.CAN_BCM).

taleinat reviewed (changes requested):
This certainly needs tests, at least checking availability as done for the other `socket.CAN_BCM_*` constants in `test_socket.BasicCANTest.testBCMConstants()`.

karlding committed:
bpo-37085: Only use #ifdef for CAN_FD_FRAME

The CAN_FD_FRAME flag was introduced in the 4.8 kernel, while the other
ones were present since SocketCAN drivers were mainlined in 2.6.25. As
such, it is probably unnecessary to guard against these constants being
missing.

karlding committed:
bpo-37085: Add test for CAN_BCM_* flag constants

karlding committed:
Add note in docs about CAN_BCM_CAN_FD_FRAME

The CAN_BCM_CAN_FD_FRAME constant mirrors the availability in the Linux
kernel. This constant was introduced in Linux 4.8 kernels.

karlding commented on Lib/test/test_socket.py:
```diff
@@ -1872,6 +1872,25 @@ def testBCMConstants(self):
+        socket.CAN_BCM_TX_RESET_MULTI_IDX
+        socket.CAN_BCM_RX_RTR_FRAME
+
+    @unittest.skipUnless(hasattr(socket, "CAN_BCM_CAN_FD_FRAME"),
```
Is it better to parse the kernel version from `platform.release` as a way of checking whether this exists?

taleinat replied:
Yes, using the `test.support.requires_linux_version` decorator.

taleinat replied:
And don't forget to also use the same decorator as above, i.e.:
```python
    @unittest.skipUnless(hasattr(socket, "CAN_BCM"),
                         'socket.CAN_BCM required for this test.')
    @requires_linux_version(4, 8)
    def testBCMConstantsSupportsFD(self):
        # ...
```

taleinat replied:
Also the method name for the separate test is unclear to me ("supports FD"). Either of `testBCMConstantsLinux48` or `testBCMConstantsFD` seems better.

karlding committed:
bpo-37085: Use support.requires_linux_version

Use support.requires_linux_version to guard BCM flag CAN_FD_FRAME,
which is only present on Linux >= 4.8.

taleinat commented on Misc/NEWS.d/next/Library/2019-06-18-16-29-31.bpo-37085.GeYaD6.rst:
```diff
@@ -0,0 +1 @@
+Added support for the Linux SocketCAN Broadcast Manager API constants.
```
Some BCM constants were already there; perhaps more precisely describe the nature of the added constants/flags.

karlding committed:
bpo-37085: Update NEWS entry

karlding committed:
bpo-37085: Remove test for CAN_BCM_CAN_FD_FRAME

This constant doesn't exist on Ubuntu 16.04 LTS, since linux-libc-dev
ships 4.4.x headers from the repos, while the feature was added in the
4.8.x kernel series. However, Ubuntu 16.04 supports the 4.15.x kernel
series, but since linux-libc-dev isn't updated past 4.4.x headers, there
is no guarantee that /usr/include/linux/can/bcm.h contains the proper
constant for the userspace API, despite having a supported kernel.

serhiy-storchaka commented on Modules/socketmodule.c:
```diff
@@ -7668,6 +7670,23 @@ PyInit__socket(void)
+    PyModule_AddIntConstant(m, "CAN_BCM_RX_ANNOUNCE_RESUME", RX_ANNOUNCE_RESUME);
+    PyModule_AddIntConstant(m, "CAN_BCM_TX_RESET_MULTI_IDX", TX_RESET_MULTI_IDX);
+    PyModule_AddIntConstant(m, "CAN_BCM_RX_RTR_FRAME", RX_RTR_FRAME);
+#ifdef  CAN_FD_FRAME
```
```suggestion
#ifdef CAN_FD_FRAME
```

serhiy-storchaka commented on Doc/library/socket.rst:
```diff
@@ -374,6 +374,9 @@ Constants
    .. availability:: Linux >= 2.6.25.
 
+   .. note::
+    The :data:`CAN_BCM_CAN_FD_FRAME` flag is only available on Linux >= 4.8.
```
```suggestion
      The :data:`CAN_BCM_CAN_FD_FRAME` flag is only available on Linux >= 4.8.
```

serhiy-storchaka reviewed (commented):
LGTM, just fix some spacing.

taleinat committed:
fix whitespace as per Serhiy's review

taleinat committed:
reword NEWS entry

karlding committed:
Fix sphinx note directive warning in socket docs

Fix the sphinx warning message that is causing the CI failure:

    Content block expected for the "note" directive; none found.

This is done by increasing the indentation to place the text in a block.

serhiy-storchaka reviewed (approved):
I think it can be merged without test for CAN_BCM_CAN_FD_FRAME.
